### DISCUSSION TITLE
LaTeX rendering with MathJax/KaTeX

### DISCUSSION THREAD

**Discussion by egormkn** [zauberzeug/nicegui#696] [Ideas / Feature Requests]
Hello! 

I would like to suggest adding math formulas support for markdown, which would be especially useful for scientific applications. `python-markdown2` doesn't seem to support LaTeX, so this might require changing the markdown implementation.

Also, it would be great to use markdown with LaTeX for basic elements such as links, buttons and checkboxes.

**Comment by al-eax** [comment]
I use markdown + latex for most of academical writings. Big plus for that propose

**Comment by rodja** [comment]
Great suggestion. We would love to see LaTeX formulas integrated. Any pull-requests towards this goal are welcome!

**Comment by sonnygeorge** [comment]
Also a big +1 here. I've built a NiceGUI app (that I use a lot) with a crucial feature that would rely on this. I will probably do some googling this weekend to see if there are any low-hanging ways to build this into NiceGUI. I will update this thread if anything seems promising. If there is an obvious path, I will try and submit a PR if I can find the time.

**Comment by sonnygeorge** [comment]
Here is a quick hack for anybody in the meantime. It uses [this](https://pypi.org/project/Markdown/) markdown implementation with [pymdown](https://pypi.org/project/pymdown-extensions/) extensions.

```python
import re

from markdown import Markdown
from nicegui import ui

HIGHLIGHT_STYLE = "background: #f5f5f5; border-radius: 0.2rem;"
HIGHLIGHT_STYLE += "padding: 0.2rem 0.3rem 0.2rem 0.3rem;"


def markdown_to_html_with_math(markdown_text: str) -> str:
    md_parser = Markdown(
        extensions=[
            "pymdownx.superfences",
            "pymdownx.highlight",
            "pymdownx.arithmatex",
            "pymdownx.inlinehilite",
        ],
        extension_configs={
            "pymdownx.inlinehilite": {
                "style_plain_text": True,
            },
            "pymdownx.highlight": {
                "guess_lang": False,
                "noclasses": True,
            },
            "pymdownx.arithmatex": {
                "smart_dollar": False,
                "preview": True,
                "generic": True,
            },
        },
    )
    return md_parser.convert(markdown_text)


def apply_tailwind(html: str) -> str:  # Borrowed from NiceGUI's ui.markdown
    rep = {
        "<h1": '<h1 class="text-5xl mb-4 mt-6"',
        "<h2": '<h2 class="text-4xl mb-3 mt-5"',
        "<h3": '<h3 class="text-3xl mb-2 mt-4"',
        "<h4": '<h4 class="text-2xl mb-1 mt-3"',
        "<h5": '<h5 class="text-1xl mb-0.5 mt-2"',
        "<a": '<a class="underline text-blue-600 hover:text-blue-800 visited:text-purple-600"',  # noqa: E501
        "<ul": '<ul class="list-disc ml-6"',
        "<p>": '<p class="mb-2">',
    }
    pattern = re.compile("|".join(rep.keys()))
    return pattern.sub(lambda m: rep[re.escape(m.group(0))], html)


def apply_custom_highlight_style(html: str) -> str:  # Would be better w/ re
    classes_to_highlight = ["highlight"]
    for class_name in classes_to_highlight:
        html = html.replace(
            f'class="{class_name}"',
            f'class="{class_name}" style="{HIGHLIGHT_STYLE}"',
        )
    return html


def markdown(text: str) -> ui.html:
    html = markdown_to_html_with_math(text)
    html = apply_tailwind(html)
    html = apply_custom_highlight_style(html)
    return ui.html(html)


if __name__ in {"__main__", "__mp_main__"}:
    # The presence of these scripts in page body is necessary for MathJax to render
    mathjax_scripts = """
    <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
    <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js"></script>
    """
    ui.add_body_html(mathjax_scripts)

    # Raw string is important for parser to not remove '\b', '\t', etc. from math
    example_md_text = r"""
This is inline code: `hello world`

This is block code:

(add your own since putting backticks here would confuse GitHub)

This is inline math: $\text{det}(A)$

This is block math:

$$ A = \begin{bmatrix} a & b \\ c & d \end{bmatrix} $$

This is a gif:

![silly cat](https://media.giphy.com/media/vFKqnCdLPNOKc/giphy.gif)
"""

    markdown(example_md_text)
    ui.run()
```

**Reply by retsyo** [reply]
I have 
```
pip install pymdown-extensions
```
however the above code does not render math yet :(

**Reply by retsyo** [reply]
well it turns out that my isp does not allow me to use `https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js`. So I host it on my local disk. Now @sonnygeorge 's code works

Then I can use `https://github.com/oruelle/md_mermaid` Extensions according to https://github.com/oruelle/md_mermaid/issues/9

Now there is only one missing feature to meet all my disire on `markdown` element, i.e. orthodox flowchart with only horizontal and vertical lines only

| feature               | support in @sonnygeorge 's markdown solution?                                                  |
| --------------------- | --------------------------------------------------------- |
| heading               | yes                                                       |
| in-line formula       | yes                                                       |
| block formula         | yes                                                       |
| table                 | yes                                                       |
| code highlight        | yes                                                       |
| pie chart via mermaid | yes<br>I know nicegui support matplotlib                  |
| flowchart via mermaid | yes and no<br>Mermaid uses curve instead of straight line |

**Comment by frankvp11** [comment]
I wonder if something like this would work - and have it be it's own element (temporarily ofc) 
https://vue-mathjax.vercel.app/guide/quick-start.html
Obviously, It'd be nice if it could be supported in the markdown, but perhaps a new element might be required?

**Comment by stepheweffie** [comment]
```py
from nicegui import ui


class LaTeXTemplate(ui.element):
    def __init__(self, script: str, template: str):
        super().__init__()
        self.script = script
        self.template = template
        ui.add_head_html(self.script)
        ui.html(self.template)
head_html = '''
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script type="text/javascript" id="MathJax-script" async
  src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js">
</script>
'''

body_html = '''
<body>
When \(a \ne 0\), there are two solutions to \(ax^2 + bx + c = 0\) and they are
$$x = {-b \pm \sqrt{b^2-4ac} \over 2a}.$$
</body>
'''
custom_index_template = True


@ui.page(path='/')
def index():

    if custom_index_template is True:
        with ui.row():
            LaTeXTemplate(head_html, body_html)


ui.run()
```

Renders beautifully 
<img width="1440" alt="Screen Shot 2023-12-28 at 8 14 39 PM" src="https://github.com/zauberzeug/nicegui/assets/1074503/4394c3fc-8782-4e2c-8329-b8c2f25157ce">
[https://docs.mathjax.org/en/latest/basic/mathematics.html#basic-mathematics](url)

**Reply by stepheweffie** [reply]
If you use the MathJax head script, adding to the page the [ui.markdown with code](https://nicegui.io/documentation/markdown#markdown_with_code_blocks) will break it. I don't know what else breaks it rn. 
Instead pass as an HTML code block and figure out your own highlighter.

```py
strCode = '''
<code class="language-python">
from nicegui import ui
<br><br>
    ui.label('Hello World!')
<br><br>
    ui.run(dark=True)
<code>
'''
```

**Reply by frankvp11** [reply]
Hmm...
I was trying to use this for a personal project wherein it would be called via a button press, but for some reason this doesn't work.
I've tried adding the ui.add_head_html to the top of the page, but that doesn't fix the issue either. Currently I get no console or terminal errors (for some reason). 
I've attached my code below:
```py
from nicegui import ui


class LaTeXTemplate(ui.element):
    def __init__(self, script: str, template: str):
        super().__init__()
        self.script = script
        self.template = template
        ui.add_head_html(self.script)
        ui.html(self.template)
head_html = '''
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script type="text/javascript" id="MathJax-script" async
  src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js">
</script>
'''
ui.add_head_html(head_html)


body_html = '''
<body>
When \(a \ne 0\), there are two solutions to \(ax^2 + bx + c = 0\) and they are
$$x = {-b \pm \sqrt{b^2-4ac} \over 2a}.$$
</body>
'''


def click_handler():
    with ui.row().style("background-color:lightblue;"):
        LaTeXTemplate(head_html, body_html)
        
ui.button("Click me", on_click=click_handler)
ui.run()
```
Note: Just running the function click_handler() does render it properly.
Edit: I need this to work since the equation isn't known until later - I've attached an example below where you can uncomment the 2 lines with input and button. Essentially, the user would input the equation and it would render. For some reason, whenever I have anything else on the page (ie the button and input) the equation doesn't get rendered properly.
```py
from nicegui import ui

head_html = '''
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script type="text/javascript" id="MathJax-script" async
  src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js">
</script>
'''
ui.add_head_html(head_html)

def click_handler(equation):
    with ui.card().style("background-color:lightblue;"):
        with ui.element("p"):
            ui.html(f'''
            <body> $${equation}$$ </body>
            ''')


click_handler("3x^2+7x+3")
# inputter = ui.input("Enter equation",  value="3x^2+7x+2")
# ui.button("Render", on_click=lambda : click_handler(inputter.value))

ui.run()
```

**Reply by stepheweffie** [reply]
You have to put it in after a page decorator 

```py
@ui.page(path='/')
def index():
    with ui.row():
        LaTeXTemplate(head_html, body_html)
    with ui.row():
        ui.markdown(strTab)
    with ui.row():
        ui.html(strCode)

```

**Comment by retsyo** [comment]
Based on above code supplied by sonnygeorge, I get a more universal Markdown element. You can find it [here](https://github.com/retsyo/study_NiceGUI)


TL;DR
I am too lazy to split markdown text into various elements. What I am dreaming is a universal Markdown element which can support as many features as https://support.typora.io/ does. But at least the following features

| feature               | support?                                                     |
| --------------------- | ------------------------------------------------------------ |
| heading               | yes                                                          |
| in-line formula       | yes                                                          |
| block formula         | yes                                                          |
| table                 | yes                                                          |
| code highlight        | yes                                                          |
| pie chart via mermaid | yes<br> I know nicegui support matplotlib                    |
| flowchart via mermaid | yes but not a good one<br>Mermaid uses curve instead of straight line |

the following code is what I get currently. Please note
- I use JavaScripts which is stored on my local disk due to my internet connection problem
- the official `md_mermaid` extension have 2 bugs
  - it deletes all unicode char
  - it inserts html head later, which is not supported by NiceGUI
- if I comment out the call of `ui.markdown`, some of the `mermaid` strings can not be plotted in my code
- the code uses `~~~mermaid ... ~~~` to define `mermaid` block

**Reply by falkoschindler** [reply]
I just noticed that it is missing in the documentation, but NiceGUI's `ui.markdown` also supports Mermaid:

```py
ui.markdown('''
    Mermaid:

    ```mermaid
    graph TD;
        Node_A --> Node_B;
    ```
''', extras=['mermaid', 'fenced-code-blocks'])
```

**Reply by retsyo** [reply]
I wish the supporting of formula is missing from document

**Comment by retsyo** [comment]
it is tedious to treat markdown files, is there any markdown editor widget fro VUE which 
- supports `table`, `code fence`, `mermaid`, `Latex formula` extended features
- can render preview window only
- can hide any toolbar, menu

if we can find one, I think we can pretend it to be the NiceGUI markdown component

**Reply by T2XX** [reply]
Good idea, I have written a wheel following this idea, which may be helpful to you
see:[nicegui_super_extensions](https://github.com/T2XX/nicegui_super_extensions)

**Comment by ryan-satterlee** [comment]
https://github.com/trentm/python-markdown2/pull/573

**Reply by falkoschindler** [reply]
Interesting! The markdown2 package seems to have added a "latex" extra in version 2.5. It requires the latex2mathml package though.

Example:
```py
ui.markdown('$e^{i\pi} = -1$', extras=['latex'])
```

<img width="72" alt="Screenshot 2024-10-11 at 07 30 07" src="https://github.com/user-attachments/assets/565f1af5-1da1-4894-b54a-8ebf80d7d083">

We should add a demo to the documentation.

**Reply by falkoschindler** [reply]
See PR #3860.

**Comment by falkoschindler** [comment]
PR #3860 just got released in version 2.4.0, implementing LaTeX support in `ui.markdown`.
Therefore I'll close this feature request as completed. ✔️
Thanks to everyone who contributed to this solution!
